### PR TITLE
Draft of PS Lite to BigTable perf test

### PR DESCRIPTION
**Pull request by pabloem**

### CODE REVIEW

**Comment by pabloem** [comment]
r: @manavgarg can you please take a look at the test implementation? : )

this is the first load test for syndeo. The medium configuration test works well. The large configurtaion does not yet work well. I am debugging and improving.

**Comment by pabloem** [comment]
also r: @pranavbhandari24 please review on usage of our testing frameworks and facilities : )

File: syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoLoadLT.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+                            "keyColumns",
+                            List.of("sha1")))));
+
+    SyndeoTemplate.buildPipeline(syndeoPipeline, SyndeoTemplate.buildFromJsonPayload(jsonPayload));
+    PipelineResult syndeoResult =
+        syndeoPipeline.runWithAdditionalOptionArgs(
+            List.of("--runner=" + testConfig.getRunner(), "--experiments=enable_streaming_engine"));
+    generatorResult.waitUntilFinish();
+
+    // TODO(pabloem): Handle cancelling the Syndeo Pipeline based on throughput/processing metrics.
```
Comment (manavgarg) [inline syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoLoadLT.java:223]: do we also need to check for some results/metrics to determine if the test ran successfully as expected?

**Review by manavgarg — DISMISSED** [review]
overall, looks good to me. Did we actually test it with dataflow runner?

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+  /** Delete any topics or subscriptions created by this manager. */
+  void cleanupAll();
+
+  /** Default exception type for {@link PubsubLiteResourceManager}. */
+  class PubsubliteResourceManagerException extends Exception {}
+
+  /** Default implementation for {@link PubsubLiteResourceManager}. */
+  class DefaultPubsubliteResourceManager implements PubsubLiteResourceManager {
+
```
Comment (pranavbhandari24) [inline it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java:82]: nit: Can you move both of those to separate files, similar to other resource managers?

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+   *
+   * @param topicName Topic name to create. The underlying implementation may not use the topic name
+   *     directly, and can add a prefix or a suffix to identify specific executions.
+   * @return The instance of the TopicName that was just created.
+   */
+  TopicName createTopic(String topicName, ReservationPath reservationPath);
+
+  /**
+   * Creates a subscription at the specific topic, with a given name.
+   *
```
Comment (pranavbhandari24) [inline it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java:64]: nit: reservationPath not in javadoc

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+   * @return
+   */
+  ReservationPath createReservation(
+      String reservationName, String cloudRegion, String projectId, Long capacity);
+
+  /**
+   * Creates a topic with the given name on Pub/Sub.
+   *
+   * <p>https://cloud.google.com/pubsub/lite/docs/reservations
+   *
```
Comment (pranavbhandari24) [inline it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java:55]: nit: reservationPath not in javadoc

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                .build();
+        Topic topic =
+            client
+                .createTopic(
+                    Topic.newBuilder()
+                        .setName(topicPath.toString())
+                        .setPartitionConfig(
+                            Topic.PartitionConfig.newBuilder()
+                                .setCount(
+                                    100) // TODO(pabloem): Define a certain number of partitions(?)
```
Comment (pranavbhandari24) [inline it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java:135]: nit: maybe add a constant

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+                                .build())
+                        .setRetentionConfig(
+                            Topic.RetentionConfig.newBuilder()
+                                // TODO(pableom): Define proper constant for period of retention.
```
Comment (pranavbhandari24) [inline it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java:139]: nit: as the todo says, add a constant?

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+                                // 5 days
+                                .setPeriod(Duration.newBuilder().setSeconds(3600 * 24 * 5).build())
+                                .setPerPartitionBytes(30 * 1024 * 1024 * 1024L)
```
Comment (pranavbhandari24) [inline it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java:142]: nit: add constant, and a comment to explain?

File: syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoLoadTestUtils.java [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+  private static <T> T generateOrNull(
+      SerializableFunction<Void, T> generator, double nullProbability, Random randomSeed) {
+    if (randomSeed.nextDouble() > nullProbability) {
+      return generator.apply(null);
+    } else {
+      return null;
+    }
+  }
+
+  public static Row randomRowForSchema(
```
Comment (pranavbhandari24) [inline syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoLoadTestUtils.java:123]: You could use the StreamingDataGenerator here. It allows you to generate data according to a user specified schema to a sink at a specified QPS. You can add the pubsublite sink to the template and use that here. 

That will reduce a lot of the code here. I have used that in all the performance tests for templates.

**Review by pranavbhandari24 — APPROVED** [review]
Looks great! Few nitpicks and suggestions, feel free to address the ones that make sense

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java (hunk 7 above)
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java]: done

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java (hunk 6 above)
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java]: done

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java (hunk 7 above)
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java]: one

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java (hunk 5 above)
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java]: done

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java (hunk 2 above)
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java]: done

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java (hunk 4 above)
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java:42]: done

File: it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java (hunk 3 above)
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsublite/PubsubLiteResourceManager.java:52]: done

File: .github/workflows/java-pr.yml [hunk 9]
```diff
@@ -77,7 +77,7 @@ jobs:
   java_integration_tests_syndeo:
     name: Syndeo Integration Tests
```
[the remark this answers is not in the document]
Comment (pabloem) [reply .github/workflows/java-pr.yml]: done. I've split the workflows. Hopefully this eases the burden on others! : )

File: syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoLoadLT.java (hunk 1 above)
Comment (pabloem) [reply syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoLoadLT.java]: I've added this check. Thanks.

And yes, this has run on dataflow

**Pull request merged**